File: app/security.py
"""Hashowanie haseł użytkowników portalu (bez dodatkowych zależności)."""

from __future__ import annotations

import hashlib
import hmac
import secrets

_ITERATIONS = 390_000
_SALT_BYTES = 16


def hash_password(password: str) -> str:
    salt = secrets.token_bytes(_SALT_BYTES)
    dk = hashlib.pbkdf2_hmac("sha256", password.encode("utf-8"), salt, _ITERATIONS)
    return f"pbkdf2_sha256${_ITERATIONS}${salt.hex()}${dk.hex()}"


def verify_password(password: str, stored: str) -> bool:
    try:
        _algo, it_s, salt_hex, hash_hex = stored.split("$", 3)
        iterations = int(it_s)
        salt = bytes.fromhex(salt_hex)
        expected = bytes.fromhex(hash_hex)
    except (ValueError, TypeError):
        return False
    dk = hashlib.pbkdf2_hmac("sha256", password.encode("utf-8"), salt, iterations)
    return hmac.compare_digest(dk, expected)

File: crm-portal-ts/src/security.ts
import { pbkdf2Sync, randomBytes, timingSafeEqual } from "node:crypto";

const ITERATIONS = 390_000;
const SALT_BYTES = 16;

export function hashPassword(password: string): string {
    const salt = randomBytes(SALT_BYTES);
    const derivedKey = pbkdf2Sync(password, salt, ITERATIONS, 32, "sha256");
    return `pbkdf2_sha256$${ITERATIONS}$${salt.toString("hex")}$${derivedKey.toString("hex")}`;
}

export function verifyPassword(password: string, stored: string): boolean {
    try {
        const [algorithm, iterationsString, saltHex, hashHex] = stored.split("$", 4);
        if (algorithm !== "pbkdf2_sha256") {
            return false;
        }

        const iterations = Number.parseInt(iterationsString, 10);
        const salt = Buffer.from(saltHex, "hex");
        const expected = Buffer.from(hashHex, "hex");
        const derivedKey = pbkdf2Sync(password, salt, iterations, expected.length, "sha256");
        return timingSafeEqual(derivedKey, expected);
    } catch (_error) {
        return false;
    }
}
